contrib/insights/src/column_analyzer.cxx

```diff
@@ -0,0 +1,120 @@
+ */
+
+#include "mlio.h"
+#include "mlio/parser.h"
+#include "mlio/util/number.h"
+#include "mlio/util/string.h"
+#include "tbb/tbb.h"
+
+#include "column_analyzer.h"
+
+#include <exception>
+#include <functional>
+#include <string>
+#include <cmath>
+#include <limits>
+#include <map>
+#include <utility>
+
+
+using namespace tbb;
+
+namespace insights {
+namespace detail {
+
+
+void column_analyzer::operator()(const blocked_range<size_t>& r) const {
+    for (size_t feature_idx = r.begin(); feature_idx != r.end(); ++feature_idx) {
+        auto t = (*current_example)->features().at(feature_idx);
+        column_analysis & feature_statistics = columns->at(feature_idx);
+        
+        auto dt = static_cast<mlio::dense_tensor*>(t.get());
+        auto cells = dt->data().as<std::string>();
+
+        auto should_capture = capture_columns->count(feature_idx) > 0;
+
+        // Used to update the mean.
+        // We do this here to aggregate several entries together and avoid
+        // potential numerical problems updating the mean a single entry at a time. 
+        double numeric_column_sum = 0.0;
+        double numeric_column_count = 0.0;
+
+        for (std::string as_string : cells) {
+            feature_statistics.rows_seen += 1;
+            feature_statistics.example_value = as_string;
+            
+            if (should_capture) {
+                if (feature_statistics.string_captured_unique_values.size() < max_capture_count) {
+                    feature_statistics.string_captured_unique_values.insert(as_string);
+                } else if (feature_statistics.string_captured_unique_values.count(as_string) == 0) {
+                    // If the value isn't present but we're not adding it because we're at a limit
+                    // then we should flag that we have overflowed.
+                    feature_statistics.string_captured_unique_values_overflowed = true;
+                }
+            }
+
+            // String analyzers.
+            feature_statistics.string_cardinality_estimator.add(as_string);
+            
+            feature_statistics.string_min_length = 
+                std::min(feature_statistics.string_min_length, as_string.size());
+            
+            feature_statistics.string_max_length = 
+                std::max(feature_statistics.string_max_length, as_string.size());
+
+            if (as_string.size() == 0) {
+                feature_statistics.string_empty_count += 1;
+            } else {
+                feature_statistics.string_min_length_not_empty = 
+                    std::min(feature_statistics.string_min_length_not_empty, as_string.length());
+            }
+
+            if (mlio::only_whitespace(as_string)) {
+                feature_statistics.string_only_whitespace_count += 1;
+            }
+
+            auto lowercase_string = as_string;
+            std::transform(as_string.begin(), as_string.end(), lowercase_string.begin(), ::tolower);
```

Comment:
1) Use `std::tolower` instead of `::tolower`.

2) Note that `tolower` is locale specific and cannot handle Unicode characters for which upper-/lower-case operations are ambigious (e.g. upper-case German ß to lower-case "ss" or ß.) As long as we assume that user-provided NaN values are within the 7-bit ASCII range, we are safe; otherwise, this needs revision.

3) Note that you are allocating a new string with `auto lowercase_string = as_string`. Not sure if this would be necessary if `mlio::matches()` was implemented in a case-insensitive way. You are effectively copying the whole dataset cell-by-cell with this statement.